imageloading-core/src/main/java/com/google/accompanist/imageloading/AsyncImage.kt

```diff
@@ -0,0 +1,367 @@
+        // that using an Image and return
+        Image(
+            painter = painterResource(previewPlaceholder),
+            contentDescription = null,
+            modifier = modifier,
+        )
+        return
+    }
+
+    val coroutineScope = rememberCoroutineScope()
+
+    val imageLoader = remember(request, coroutineScope) {
+        ImageLoader(
+            state = request,
+            coroutineScope = coroutineScope,
+            shouldRefetchOnSizeChange = shouldRefetchOnSizeChange,
+        )
+    }
+
+    // NOTE: All of the things that we want to be able to change without recreating the whole object
+    // we want to do inside of here.
+    SideEffect {
+        imageLoader.shouldRefetchOnSizeChange = shouldRefetchOnSizeChange
+    }
+
+    Box(
+        propagateMinConstraints = true,
+        modifier = modifier.then(imageLoader.layoutModifier)
+    ) {
+        content(request.loadState)
+    }
+}
+
+@Stable
+private class ImageLoader<R : Any>(
+    val state: AsyncImageState<R>,
+    val coroutineScope: CoroutineScope,
+    shouldRefetchOnSizeChange: (currentResult: ImageLoadState, size: IntSize) -> Boolean,
+) : RememberObserver {
+    // Our size to use when performing the image load request
+    var requestSize by mutableStateOf<IntSize?>(null)
+
+    // Our updated shouldRefetchOnSizeChange
+    var shouldRefetchOnSizeChange by mutableStateOf(shouldRefetchOnSizeChange)
+
+    // Current request job
+    private var job: Job? = null
+
+    // Our painter. We use a sub-class of [PainterModifier] which delegates the parameter
+    // values to be state reads. The properties will be read during layout and drawing, which means
+    // that any values changes will automatically re-trigger layout/draw as necessary.
+    // This allows us to avoid needing to recreate the modifier if any of the values change, which
+    // is what would need to happen if we used Modifier.paint().
+    val painter: Painter = DelegatingPainter {
+        state.loadState.let { state ->
+            when (state) {
+                is ImageLoadState.Success -> state.painter
+                is ImageLoadState.Error -> state.painter ?: EmptyPainter
+                else -> EmptyPainter
+            }
+        }
+    }
+
+    // Our layout modifier, which allows us to receive the incoming constraints to update
+    // requestSize. Using a modifier allows us to avoid using BoxWithConstraints and the cost of
+    // subcomposition. For most usages subcomposition is fine, but AsyncImage's tend to be used
+    // in large quantities which multiplies the cost.
+    val layoutModifier = Modifier.layout { measurable, constraints ->
+        val newSize = IntSize(
+            width = if (constraints.hasBoundedWidth) constraints.maxWidth else -1,
+            height = if (constraints.hasBoundedHeight) constraints.maxHeight else -1
+        )
+
+        if (requestSize == null ||
```

Comment:
here we can optimize more. 
so now you read a requestSize state inside the measure block. this means when you change it this will schedule one more remeasure unfortunately. 
will be better to never read it and just do here:
requestSize = newSize

then the build in into mutableStateOf equals check will ignore the updates to the same value.
then where we listen to it 
snapshotFlow { requestSize }
we can filter when shouldRefetchOnSizeChange returned false